rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp

```diff
@@ -66,18 +66,57 @@ class TestNodeGraph : public ::testing::Test
 
   const rclcpp::node_interfaces::NodeGraph * node_graph() const {return node_graph_;}
 
+  size_t get_num_topics()
+  {
+    constexpr std::chrono::milliseconds timeout(100);
+
+    int tries = 0;
+    size_t num_topics = 0;
+    while (num_topics < 1u && tries++ < 5) {
+      auto event = node()->get_graph_event();
+      EXPECT_NO_THROW(node()->wait_for_graph_change(event, timeout));
+
+      auto topic_names_and_types = node_graph()->get_topic_names_and_types();
+      num_topics = topic_names_and_types.size();
+    }
+
+    // Run notify_graph_change() here to remove the graph_event users we created
+    // in the loop above.
+    node_graph_->notify_graph_change();
```

Comment:
@clalancette I'm not sure I follow this bit -- or at least I find it counter-intuitive.

Reply:
>  I'm not sure I follow this bit -- or at least I find it counter-intuitive.

I agree that it is counter-intuitive, though it is the way current `NodeGraph` API is setup.

A call to [node->get_graph_event](https://github.com/ros2/rclcpp/blob/b9ef1fedfa409c966658a747d653f84cce7357c3/rclcpp/src/rclcpp/node.cpp#L386) ends up calling [node_graph->get_graph_event](https://github.com/ros2/rclcpp/blob/b9ef1fedfa409c966658a747d653f84cce7357c3/rclcpp/src/rclcpp/node_interfaces/node_graph.cpp#L368).  There, a new shared point to `rclcpp::Event` is created, a weak reference to that `Event` is stored, the `graph_users_count_` is incremented, and the shared pointer is returned.  However, there is no explicit decrement of `graph_users_count_`.  Instead, when `notify_graph_change` is called, it loops through the weak pointers, attempting to lock them.  If it finds one that it cannot lock anymore (because it has been destroyed), then it removes that weak pointer from the list and recalculates the size of `graph_users_count`.  That way you'll get the "correct" number of events when you later call `count_graph_users`.

Reply:
I see. It's out of scope for this PR, but it'd be nice to have a way to explicitly drop the event instead of doing this.

Reply:
This also seems like an abuse of this function to me. If you need to prune old events then maybe we should have a function for that, but why does that matter?

Reply:
To be specific, I confused why clearing the old events is important.

Reply:
> To be specific, I confused why clearing the old events is important.

Mostly because of this check: https://github.com/ros2/rclcpp/blob/b9ef1fedfa409c966658a747d653f84cce7357c3/rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp#L94 .  At that point, we can't really know how many graph events it took to get the previous data, so we really need to clean up.  Alternatively, we could remove that check, but I feel like it is a good thing to test.

Reply:
I don't even understand what that check is doing? Maybe just getting coverage on that function? The value it returns isn't really important I think.

I mean if we ever added something to the node that happen to use the graph events, this would break because there would be more users. I think that change could just be changed to >= 1, if needed?

Reply:
> I mean if we ever added something to the node that happen to use the graph events, this would break because there would be more users. I think that change could just be changed to >= 1, if needed?

That's true, we could do that.

Here's what I'm going to do.  To take advantage of the CI that already successfully completed, I'm going to merge this PR as-is.  That fixes one of our current problems in https://github.com/ros2/rmw/pull/293 .  I'll do an immediate follow-up to change that fragile `count_graph_users` test and remove the `notify_graph_change`, which should satisfy @wjwwood 's concerns.